### ISSUE TITLE
Browser tab crashes while displaying an heavy map (400 style layers)

### ISSUE THREAD

**Issue by jean-kiepura** [maplibre/maplibre-gl-js#3023]
We have a pretty heavy Map (about 400 style layers), mainly made of vector layer, with a couple of raster layers.

We recently ported our application  from MapBox GL JS v1 to MapLibre v3, with great success.

However, we are facing performance problems and crashes while loading this particular project that contains 400 layers.

Unfortunately, the Map data is confidential and I can't really provide a link to that map.

**maplibre-gl-js version**: ^3.3.0

**browser**: Chrome, Firefox (Mac OS)

When we display the same project with MapBox, the Browser tab eats up to 1.3 gigs of memory but does not crash.

Using MapLibre, here are the numbers we get:

 - Chrome: memory footprint goes up to around 4.5 gigs and the browser tab crashes
 - Firefox: memory footprint goes up to  18 gigs, the browser tab does *not* crash but this memory footprint is clearly unreasonable

I have provided screen shots of the browsers task managers.

Is is good to note that most of the 400 style layers are initially not visible, but the end user can turn them on and off via the app GUI. Chromes crashes at startup while most of layers are not visible on the map.

I have set the number of web workers in maplibregl to half of the computer available cores, but this doesn't help getting the memory footprint down.

In Chrome, at some point, the browser stops execution before an OOM condition. At that point, MapLibre seems to be in the process of rendering features on the map (see screenshot).

My first impression would be to suspect some kind of garbage collection issue.

Chrome with Mapbox GL JS v1:
![image](https://github.com/maplibre/maplibre-gl-js/assets/44607784/5cc881ae-5b2f-4bf8-80f3-c281ebc7928d)

Firefox with MapLibre:
![image](https://github.com/maplibre/maplibre-gl-js/assets/44607784/0f776b76-a8fe-4ac6-9045-b6bd8b806b0b)

Chrome with MapLibre:
![image](https://github.com/maplibre/maplibre-gl-js/assets/44607784/c7a45233-e9b1-435a-a377-c2ea300dadd4)

Chrome debugger stopping before OOM
![image](https://github.com/maplibre/maplibre-gl-js/assets/44607784/cf59a0a7-9220-4373-9f73-3a88b3588289)

### Expected Behavior
Browser loads the map with a reasonable amount of memory, and doesn't crash

### Actual Behavior
Browser eats up to 18 gigs of memory (Firefox) , or crashes before map loading is finished (Chrome)

**Comment by HarelM** [comment]
Since the data is confidential I think your best bet would be to debug this yourself and see what's eating up the memory.
Another approach would be to try versions between 1.x and 3.x and see in which version this starts to happen.
We have made sure to release a lot of versions so that narrowing down a problematic feature that was added will be possible.
If this still reproduces in version 1.x then I'm not sure what you described is correct as 1.x is equal with Mapbox's 1.x version.

**Comment by jean-kiepura** [comment]
I have landmarks in our progressions for ML v1 and V2, I could for sure try that and get back with details. (our own app had to be refactored for ML versions)

Would it be useful to provide you with the style JSON? but I would have to edit a lot of sensible data (names, source-layers, filters, source URLs, etc); it would be a bit long and maybe not that useful, but if you tell me that it could help I could prepare that.

**Comment by HarelM** [comment]
If you could make a reproducible example in jsbin that shows that the memory changed dramatically between version 1 and 3 it can help, other shortcuts won't do, sorry...

**Comment by jean-kiepura** [comment]
@HarelM , during the process of preparing a public access to our data, I discovered that the production code of our application (minified code, including MapLibre, hosted on a CDN) is not affected by the memory usage behaviour.

So I guess that the high-memory usage is related to our dev environnement (when I run our app locally through a dev server), and since we had to upgrade some part of our tech setup when we upgraded to MapLibre, it's not very clear who's the culprit.

In "production" our project loads very fast using MapLibre 3 and stay stable at 450 megs of memory (versus 3.5 gigs or 14 gigs in dev with Chrome and Firefox).

So I guess we can close this ticket, thanks a lot for the fast response. Up to now, we are very happy of having decided to switch to MapLibre!
